R/http-libcurl.R

```diff
@@ -87,75 +44,101 @@ httpLibCurl <- function(protocol,
       upload = TRUE,
       infilesize_large = fileLength,
       readfunction = function(nbytes, ...) {
-        if (is.null(con)) {
+        if (is.null(contentCon)) {
           return(raw())
         }
-        bin <- readBin(con, "raw", nbytes)
+        bin <- readBin(contentCon, "raw", nbytes)
         if (length(bin) < nbytes) {
-          close(con)
-          con <<- NULL
+          close(contentCon)
+          contentCon <<- NULL
         }
         bin
       }
     )
   }
 
-  # ensure we're using the requested method
-  curl::handle_setopt(handle, customrequest = method)
-
-  # apply all our accumulated headers
+  headers <- appendCookieHeaders(request, headers)
   curl::handle_setheaders(handle, .list = headers)
 
   # make the request
-  response <- NULL
+  url <- buildHttpUrl(request)
   time <- system.time(
     response <- curl::curl_fetch_memory(url, handle = handle),
     gcFirst = FALSE
   )
   httpTrace(method, path, time)
 
-  # get list of HTTP response headers
+  # Process headers
   headers <- curl::parse_headers_list(rawToChar(response$headers))
 
-  # read normalized location header
-  if ("location" %in% names(headers))
-    location <- headers[["location"]]
-  else
-    location <- NULL
+  # Parse cookies from header; bear in mind that there may be multiple headers
+  cookieHeaders <- headers[names(headers) == "set-cookie"]
+  storeCookies(request, cookieHeaders)
 
   # presume a plain text response unless specified otherwise
-  contentType <- if ("content-type" %in% names(headers)) {
-    headers[["content-type"]]
-  } else {
-    "text/plain"
-  }
+  contentType <- headers[["content-type"]] %||% "text/plain"
+  contentValue <- rawToChar(response$content)
 
   # emit JSON trace if requested
-  jsonTracingEnabled <-
-    httpTraceJson() &&
-    !is.null(contentFile) &&
-    identical(contentType, "application/json")
+  jsonTracingEnabled <- httpTraceJson() && contentType == "application/json"
```

Comment:
nit: This could be a tracing function passed by the caller that receives the content-type and payload and hides the logging decisions from this function.

```r
traceResponse(contentType, contentFile, contentValue)
```

Reply:
I think the intent here is to offer something that the end user can turn on/off in case they need extra debug info.

Reply:
Sorry. I wasn't clear: I was suggesting that we construct a function dynamically, similar to how we construct the `logger` in `deployApp`. This code may be too low-level to make that type of dependency-injection workflow straightforward.

```r
traceResponse <- makeTraceResponse()

makeTraceResponse <- function() {
    if conditions-to-allow-tracing and httpTraceJson() {
        function(contentType, contentFile, contentValue) {
            # do the real logging.
        }
    } else {
        function(contentType, contentFile, contentValue) {}
    }
}
```

This was a really soft suggestion. Maybe the more important piece is to extract the logging from the http function...